packages/components/src/internal/components/assay/withAssayModels.tsx

```diff
@@ -0,0 +1,286 @@
+            );
+        };
+    }
+
+    ComponentWithAssays.defaultProps = {
+        assayLoader: defaultProps?.assayLoader ?? DefaultAssayLoader,
+    };
+
+    return ComponentWithAssays;
+}
+
+/**
+ * Provides a [[withAssayModels]] wrapped component that is additionally wrapped by react-router's withRouter.
+ * This additional wrapping allows for sourcing the "assayName" property from the URL. NOTE: This is specifically
+ * configured to expect a route param called "protocol" which is expected to a (string) name of a specific assay
+ * protocol.
+ * @param ComponentToWrap: The component definition (e.g. class, function) to wrap.
+ * This will have [[InjectedAssayModel]] props injected into it when instantiated.
+ * @param defaultProps: Provide alternative "defaultProps" for this wrapped component.
+ */
+export function withAssayModelsFromLocation<Props>(
+    ComponentToWrap: ComponentType<Props & InjectedAssayModel>,
+    defaultProps?: WithAssayModelProps
+): ComponentType<Props & WithAssayModelProps & WithRouterProps> {
+    const WrappedComponent = withAssayModels<Props>(ComponentToWrap, defaultProps);
+
+    const AssayFromLocation: FC<Props & WithRouterProps> = props => {
+        return <WrappedComponent {...props} assayName={props.params?.protocol} />;
+    };
+
+    return withRouter(AssayFromLocation);
+}
+
+/**
+ * Returns a higher-order component wrapped with [[withAssayModelsFromLocation]] that provides common
+ * "page"-level handling for edge cases (e.g. loading, protocol not found, errors during loading, etc).
+ * @param ComponentToWrap: The component definition (e.g. class, function) to wrap.
+ * This will have [[InjectedAssayModel]] props injected into it.
+ * @param defaultProps: Provide alternative "defaultProps" for this wrapped component.
+ */
+export function assayPage<Props>(
+    ComponentToWrap: ComponentType<Props & InjectedAssayModel>,
+    defaultProps?: WithAssayModelProps
+): ComponentType<Props & WithAssayModelProps & WithRouterProps> {
+    const AssayPageImpl: FC<Props & InjectedAssayModel & WithRouterProps> = props => {
+        const { assayModel, params } = props;
+        const assayName = params?.protocol;
+        const hasProtocol = assayName !== undefined;
+
+        if (
+            isLoading(assayModel.definitionsLoadingState) ||
+            (hasProtocol && isLoading(assayModel.protocolLoadingState))
+        ) {
+            return <LoadingPage />;
+        }
+
+        if (assayModel.definitionsError || assayModel.protocolError) {
+            if (hasProtocol && assayModel.getByName(assayName) === undefined) {
+                return <NotFound />;
+            }
+
+            return (
+                <Alert>{getActionErrorMessage('There was a problem loading the assay design.', 'assay design')}</Alert>
+            );
+        }
+
+        return <ComponentToWrap {...props} />;
+    };
+
+    return withAssayModelsFromLocation(AssayPageImpl, defaultProps);
```

Comment:
Nice, I like the addition of these helpers (withAssayModelsFromLocation and withAssayModelsFromLocation).

Reply:
Thanks!